go/vt/automation/scheduler.go

```diff
@@ -0,0 +1,258 @@
+	// Guarded by "mu".
+	toBeScheduledClusterOperations chan *ClusterOperationInstance
+	// Guarded by "mu".
+	state schedulerState
+
+	pendingOpsWg *sync.WaitGroup
+
+	muOpList sync.Mutex
+	// Guarded by "muOpList".
+	activeClusterOperations map[string]*ClusterOperationInstance
+	// Guarded by "muOpList".
+	finishedClusterOperations map[string]*ClusterOperationInstance
+}
+
+// NewScheduler creates a new instance.
+func NewScheduler() (*Scheduler, error) {
+	defaultClusterOperations := map[string]bool{
+		"ReshardingTask": true,
+		// Testing only cluster operations.
+		"TestingEchoTask":     true,
+		"TestingEmitEchoTask": true,
+	}
+
+	s := &Scheduler{
+		registeredClusterOperations:    defaultClusterOperations,
+		idGenerator:                    IDGenerator{},
+		toBeScheduledClusterOperations: make(chan *ClusterOperationInstance, 10),
+		state:                     stateNotRunning,
+		pendingOpsWg:              &sync.WaitGroup{},
+		activeClusterOperations:   make(map[string]*ClusterOperationInstance),
+		finishedClusterOperations: make(map[string]*ClusterOperationInstance),
+	}
+
+	return s, nil
+}
+
+// Run processes queued cluster operations.
+func (s *Scheduler) Run() {
+	s.mu.Lock()
+	s.state = stateRunning
+	s.mu.Unlock()
+
+	s.pendingOpsWg.Add(1)
+	go s.processRequestsLoop()
+}
+
+func (s *Scheduler) processRequestsLoop() {
+	defer s.pendingOpsWg.Done()
+
+	for op := range s.toBeScheduledClusterOperations {
+		s.processClusterOperation(op)
+	}
+	log.Infof("Stopped processing loop for ClusterOperations.")
+}
+
+func (s *Scheduler) processClusterOperation(clusterOp *ClusterOperationInstance) {
+	if clusterOp.State == pb.ClusterOperationState_CLUSTER_OPERATION_DONE {
+		log.Infof("ClusterOperation: %v skipping because it is already done. Details: %v", clusterOp.Id, clusterOp)
+		return
+	}
+
+	log.Infof("ClusterOperation: %v running. Details: %v", clusterOp.Id, clusterOp)
+
+	var lastTaskError string
+	for i := 0; i < len(clusterOp.SerialTasks); i++ {
+		taskContainer := clusterOp.SerialTasks[i]
+		for _, taskProto := range taskContainer.ParallelTasks {
+			if taskProto.State == pb.TaskState_DONE {
+				if taskProto.Error != "" {
+					log.Errorf("Task: %v (%v/%v) failed before. Aborting the ClusterOperation. Error: %v Details: %v", taskProto.Name, clusterOp.Id, taskProto.Id, taskProto.Error, taskProto)
+					lastTaskError = taskProto.Error
+					break
+				} else {
+					log.Infof("Task: %v (%v/%v) skipped because it is already done. Full Details: %v", taskProto.Name, clusterOp.Id, taskProto.Id, taskProto)
+				}
+			}
+
+			task, err := s.createTaskInstance(taskProto.Name)
+			if err != nil {
+				log.Errorf("Task: %v (%v/%v) could not be instantiated. Error: %v Details: %v", taskProto.Name, clusterOp.Id, taskProto.Id, err, taskProto)
+				MarkTaskFailed(taskProto, err)
+				lastTaskError = err.Error()
+				break
+			}
+
+			log.Infof("Task: %v (%v/%v) running. Details: %v", taskProto.Name, clusterOp.Id, taskProto.Id, taskProto)
```

Comment:
When should tasks be marked as running? Will each task be responsible for doing that itself?